### ISSUE TITLE
27e904532594b7fb (fix for #23810) introduces a regression

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10384]
Migrated from [rt.perl.org#75146](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=75146) (status was 'resolved')

Searchable as RT75146$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



27e904532594b7fb\, the fix for "RT 23810&#8203;: eval and tied methods"\, post 5\.12\.0\,
introduces a regression when eval fails inside a code reference in @&#8203;INC\.
I'll class this as "high" because it's a double free\, and that's what
bad stuff\(TM\) is made from\.

Test case is a massively cut down version of Tangram's 05\-generator\-2\.t

$ cat \.\./tgboom\.pl 
\#\!/usr/bin/perl \-w

use strict;

unshift @&#8203;INC\, sub \{
&nbsp;    my \($self\, $fn\) = @&#8203;\_;

&nbsp;    \(my $pkg = $fn\) =~ s\{/\}\{&#8203;::\}g;
&nbsp;    $pkg =~ s\{\.pm$\}\{\};

&nbsp;    if \($pkg eq 'Credit'\) \{
&nbsp;        my $code = \<\<'EOC';
package Credit;

use NonsenseAndBalderdash;

1;
EOC
&nbsp;        eval $code;
&nbsp;        die "\\$@&#8203; is $@&#8203;";
&nbsp;    \}

&nbsp;    \#print STDERR "Generator&#8203;: not one of mine\, ignoring\\n";
&nbsp;    return undef;
\};

\# create load\-on\-demand new\(\) constructors
\{
&nbsp;    package Credit;
&nbsp;    sub new \{
&nbsp;        eval "use Credit";
&nbsp;    \}
\};

eval \{
&nbsp;    my $credit = new Credit;
\};

print "If you get here\, you didn't crash\\n";
$ \./perl \-Ilib \.\./tgboom\.pl 
\*\*\* glibc detected \*\*\* \./perl&#8203;: free\(\)&#8203;: invalid pointer&#8203;: 0x00007fff833ad770 \*\*\*
======= Backtrace&#8203;: =========
/lib/libc\.so\.6\[0x2b6c28465928\]
/lib/libc\.so\.6\(cfree\+0x76\)\[0x2b6c28467a36\]
\./perl\(Perl\_op\_free\+0x1ac\)\[0x420ab9\]
\./perl\[0x613d08\]
\./perl\(Perl\_pp\_entereval\+0xf99\)\[0x61b696\]
\./perl\(Perl\_runops\_debug\+0x153\)\[0x5000da\]
\./perl\(Perl\_call\_sv\+0x44d\)\[0x44c6eb\]
\./perl\(Perl\_pp\_require\+0x1d34\)\[0x616e6b\]
\./perl\(Perl\_runops\_debug\+0x153\)\[0x5000da\]
\./perl\(Perl\_call\_sv\+0x635\)\[0x44c8d3\]
\./perl\(Perl\_call\_list\+0x27f\)\[0x457d5c\]
\./perl\[0x4361d2\]
\./perl\(Perl\_newATTRSUB\+0x33ee\)\[0x435f65\]
\./perl\(Perl\_utilize\+0x333\)\[0x42bdf0\]
\./perl\(Perl\_yyparse\+0x16dc\)\[0x4b26e8\]
\./perl\[0x613c7d\]
\./perl\(Perl\_pp\_entereval\+0xf99\)\[0x61b696\]
\./perl\(Perl\_runops\_debug\+0x153\)\[0x5000da\]
\./perl\[0x44b59a\]
\./perl\(perl\_run\+0x13a\)\[0x44a9e8\]
\./perl\(main\+0xb1\)\[0x42018d\]
/lib/libc\.so\.6\(\_\_libc\_start\_main\+0xe6\)\[0x2b6c284101a6\]
\./perl\[0x420019\]
======= Memory map&#8203;: ========
00400000\-007af000 r\-xp 00000000 08&#8203;:01 7127996                            /home/nick/Perl/perl/perl
009ae000\-009b3000 rw\-p 003ae000 08&#8203;:01 7127996                            /home/nick/Perl/perl/perl
009b3000\-009f5000 rw\-p 009b3000 00&#8203;:00 0                                  \[heap\]
2b6c276fa000\-2b6c27716000 r\-xp 00000000 08&#8203;:01 7094683                    /lib/ld\-2\.7\.so
2b6c27716000\-2b6c27719000 rw\-p 2b6c27716000 00&#8203;:00 0 
2b6c27915000\-2b6c27917000 rw\-p 0001b000 08&#8203;:01 7094683                    /lib/ld\-2\.7\.so
2b6c27917000\-2b6c2792c000 r\-xp 00000000 08&#8203;:01 7094688                    /lib/libnsl\-2\.7\.so
2b6c2792c000\-2b6c27b2b000 \-\-\-p 00015000 08&#8203;:01 7094688                    /lib/libnsl\-2\.7\.so
2b6c27b2b000\-2b6c27b2d000 rw\-p 00014000 08&#8203;:01 7094688                    /lib/libnsl\-2\.7\.so
2b6c27b2d000\-2b6c27b2f000 rw\-p 2b6c27b2d000 00&#8203;:00 0 
2b6c27b2f000\-2b6c27b31000 r\-xp 00000000 08&#8203;:01 7094681                    /lib/libdl\-2\.7\.so
2b6c27b31000\-2b6c27d31000 \-\-\-p 00002000 08&#8203;:01 7094681                    /lib/libdl\-2\.7\.so
2b6c27d31000\-2b6c27d33000 rw\-p 00002000 08&#8203;:01 7094681                    /lib/libdl\-2\.7\.so
2b6c27d33000\-2b6c27db5000 r\-xp 00000000 08&#8203;:01 7094694                    /lib/libm\-2\.7\.so
2b6c27db5000\-2b6c27fb4000 \-\-\-p 00082000 08&#8203;:01 7094694                    /lib/libm\-2\.7\.so
2b6c27fb4000\-2b6c27fb6000 rw\-p 00081000 08&#8203;:01 7094694                    /lib/libm\-2\.7\.so
2b6c27fb6000\-2b6c27fbe000 r\-xp 00000000 08&#8203;:01 7094687                    /lib/libcrypt\-2\.7\.so
2b6c27fbe000\-2b6c281be000 \-\-\-p 00008000 08&#8203;:01 7094687                    /lib/libcrypt\-2\.7\.so
2b6c281be000\-2b6c281c0000 rw\-p 00008000 08&#8203;:01 7094687                    /lib/libcrypt\-2\.7\.so
2b6c281c0000\-2b6c281ef000 rw\-p 2b6c281c0000 00&#8203;:00 0 
2b6c281ef000\-2b6c281f1000 r\-xp 00000000 08&#8203;:01 7094679                    /lib/libutil\-2\.7\.so
2b6c281f1000\-2b6c283f0000 \-\-\-p 00002000 08&#8203;:01 7094679                    /lib/libutil\-2\.7\.so
2b6c283f0000\-2b6c283f2000 rw\-p 00001000 08&#8203;:01 7094679                    /lib/libutil\-2\.7\.so
2b6c283f2000\-2b6c2853c000 r\-xp 00000000 08&#8203;:01 7094689                    /lib/libc\-2\.7\.so
2b6c2853c000\-2b6c2873b000 \-\-\-p 0014a000 08&#8203;:01 7094689                    /lib/libc\-2\.7\.so
2b6c2873b000\-2b6c2873e000 r\-\-p 00149000 08&#8203;:01 7094689                    /lib/libc\-2\.7\.so
2b6c2873e000\-2b6c28740000 rw\-p 0014c000 08&#8203;:01 7094689                    /lib/libc\-2\.7\.so
2b6c28740000\-2b6c28747000 rw\-p 2b6c28740000 00&#8203;:00 0 
2b6c28747000\-2b6c2875d000 r\-xp 00000000 08&#8203;:01 7095368                    /lib/libgcc\_s\.so\.1
2b6c2875d000\-2b6c2895d


Bisect says&#8203;:

27e904532594b7fb224bdf9a05bf3b5336b8a39e is first bad commit
commit 27e904532594b7fb224bdf9a05bf3b5336b8a39e
Author&#8203;: David Mitchell \<davem@&#8203;iabyn\.com>
Date&#8203;:   Thu Apr 8 13&#8203;:16&#8203;:56 2010 \+0100

&nbsp;    fix RT 23810&#8203;: eval and tied methods

&nbsp;    Something like the following ended up corrupted&#8203;:
&nbsp;        sub FETCH \{ eval 'BEGIN\{syntax err\}' \}
&nbsp;    The croak on error popped back the context stack etc to the EVAL pushed by
&nbsp;    entereval\, but the corresponding JUMPENV\_PUSH\(3\) unwound all the way to the
&nbsp;    outer perl\_run\, losing all the mg\_get\(\) related parts of the C stack\.

&nbsp;    It turns out that the run\-time parts of pp\_entereval were protected with
&nbsp;    a new JUMPENV level\, but the compile\-time parts weren't\. Add this\.

:100644 100644 80c7b221d7e967a6f3380d70917bd73700b16852 bbb2d1587cb197977c996ecfe8abddf4f9aa3631 M      pp\_ctl\.c
:040000 040000 7fa8480b6084b533c3eaddbbbd6834e2551f8a4e f2ec8022695df42cff76ab78c117bd44d1fb626c M      t
bisect run success

I hope that the fix is easy\. Sorry Dave\, to be the bearer of bad news\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.12.0:

Configured by nick at Mon May 17 15:55:08 BST 2010.

Summary of my perl5 (revision 5 version 12 subversion 0) configuration:
  Derived from: 27e904532594b7fb224bdf9a05bf3b5336b8a39e
  Platform:
    osname=linux, osvers=2.6.18.8-xenu, archname=x86_64-linux-ld
    uname='linux eris 2.6.18.8-xenu #1 smp sat oct 3 10:27:42 bst 2009 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-g -Uusethreads -Uuse64bitall -Duselongdouble -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.5-137-g27e9045 -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.11.5-137-g27e9045 -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='long double', nvsize=16, Off_t='off_t', lseeksize=8
    alignbytes=16, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    RC0


@INC for perl 5.12.0:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.11.5-137-g27e9045/lib/perl5/site_perl/5.12.0/x86_64-linux-ld
    /home/nick/Sandpit/snap5.9.x-v5.11.5-137-g27e9045/lib/perl5/site_perl/5.12.0
    /home/nick/Sandpit/snap5.9.x-v5.11.5-137-g27e9045/lib/perl5/5.12.0/x86_64-linux-ld
    /home/nick/Sandpit/snap5.9.x-v5.11.5-137-g27e9045/lib/perl5/5.12.0
    .


Environment for perl 5.12.0:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

fixed by c86ffc3299ed26fc3616b4d2bb0abdf2e70a0ad1
